### PR TITLE
[Messenger] Add the transports documentation

### PR DESCRIPTION
**Pull request by sroze**
Add the Messenger transports documentation. Does not aim to be exhaustive but give more details.

One little thing... we wait for this one to be merged: https://github.com/symfony/symfony/pull/27008

### CODE REVIEW

File: components/messenger/transports.rst [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+       private $toEmail;
+       private $mailer;
+
+       public function __construct(\Swift_Mailer $mailer, string $toEmail)
+       {
+           $this->mailer = $mailer;
+           $this->toEmail = $toEmail;
+       }
+
+       public function send($message)
```
Comment (yceruto) [inline components/messenger/transports.rst]: Should be updated according to the current contract, `Envelope $...` ?

File: components/messenger/transports.rst [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        }
+    }
+
+The transport object is needs to implements the ``TransportInterface`` (which simply combine
+the ``SenderInterface`` and ``ReceiverInterface``). It will look
+like this::
+
+    class YourTransport implements TransportInterface
+    {
+        public function send($message) : void
```
Comment (yceruto) [inline components/messenger/transports.rst]: Same here.

File: components/messenger/transports.rst [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+~~~~~~~~~~~~~~~
+
+Using the ``SenderInterface``, you can easily create your own message sender.
+Let's say you already have an ``ImportantAction`` message going through the
+message bus and handled by a handler. Now, you also want to send this message as
+an email.
+
+First, create your sender::
+
+    namespace App\MessageSender;
```
Comment (yceruto) [inline components/messenger/transports.rst]: We could go suggesting a better structure for Messenger stuf? maybe `App\Message\Sender` or `App\Messenger\Transport\Sender` instead?

**Review by weaverryan — CHANGES_REQUESTED** [review]

File: components/messenger/transports.rst [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+.. note:
+
+    By default, RabbitMQ uses ``guest`` as a username and ``guest`` as a password
+    and has a ``/`` vhost.
+
+Error handling
+~~~~~~~~~~~~~~
+
+If something wrong happens (i.e. an exception is thrown) while handling your message,
+the default behaviour is that your message will be "NACK" and requeued.
```
[the remark this answers is not in the document]
Comment (sroze) [reply components/messenger/transports.rst]: Not Acknowledge. Do you have any idea on how to phrase it differently? 🤷‍♂️

File: components/messenger/transports.rst (hunk 3 above)
Comment (sroze) [reply components/messenger/transports.rst]: Yep, why not 👍

File: components/messenger/transports.rst [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+      ?retry[attempts]=3
+      &retry[ttl][0]=10000
+      &retry[ttl][1]=30000
+      &retry[ttl][2]=60000
+
+In the example above, if handling the message fails, it will retry it 3 times. After
+the first failure, it will wait 10 seconds before trying it. After the 2nd failure,
+it will wait 30 seconds. After the 3rd failure, it will wait a minute. If it still
+fails, the message will be moved to a "dead queue" and you will have to manually
+handle this message.
```
[the remark this answers is not in the document]
Comment (sroze) [reply components/messenger/transports.rst]: These are to be removed as retry wasn't merged in.

**Review by sroze — COMMENTED** [review]
Will update when the envelope one is merged :)

File: components/messenger/transports.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (ktherage) [reply components/messenger/transports.rst]: Beginning with this, i'll explain this like : Not Acknowledge = not _treated_ and requeued.

**Comment by javiereguiluz** [comment]
Anyone with experience in Messenger is willing to do the final push required to finish this pull request? Thanks!

**Comment by llaakkkk** [comment]
Thank you very much for your pull request!

As part of the Symfony EU funded hackathon (https://symfony.com/blog/the-symfony-and-api-platform-hackathon-is-coming), we were able to assemble most of the core team and big contributors in one place. Our goal is to finish as many open issues and PRs as possible.

*Your commits will not be lost and you will therefore get credit for your work*. All that will happen is that your commits will be moved to a new PR where all remaining concerns will be addressed.

Without your work this would not have been possible. So thank you once again!

**Comment by sroze** [comment]
Thank you @llaakkkk 🙌

**Pull request closed without merging**